Fix building package without a GPU

aphedges opened the pull request:
Fixes #2010.

The first commit is the minimal change needed to get DeepSpeed installed. The second commit that fully removes `bf16_support` from `torch_info`. If the second commit goes too far, it can be dropped before this PR is merged. I have tested both and confirmed that they both allow me to install DeepSpeed. See <https://github.com/microsoft/DeepSpeed/issues/2010#issuecomment-1165869012> for more information about my approach.

stas00 commented:
Looks great, @aphedges! Thank you for working on fixing this!

stas00 commented on setup.py:
```diff
@@ -235,13 +234,10 @@ def create_dir_symlink(src, dest):
     hip_version = ".".join(torch.version.hip.split('.')[:2])
 torch_info = {
     "version": torch_version,
-    "bf16_support": bf16_support,
```
as long as there is nothing relying on this field it should be safe to remove.

aphedges replied:
In this repository, It was only used by a test, and I've replaced the usage in the test with a runtime check.

The usage in the test was only introduced 3 weeks ago in 7fc3065074885b0a2ec59c9c49e04296dd0e0fcd, so it probably hasn't been used much by other programs.

stas00 replied:
I totally agree, but let's see what the maintainers say.

stas00 replied:
as you pointed out the key here is that the hardware environment may chance since the installation and thus a memorized bf16 state can easily be invalid at run time. 

This would be especially true for cross-platform/cross-hardware prebuilds.

stas00 replied:
@jeffra, also there is a new development with IPEX - which now supports BF16 on CPU! So I highly recommend to already split up into 2 different state checks - have_bf16_cpu and have_bf16_gpu as we have just went through in transformers. https://github.com/huggingface/transformers/pull/17738

mrwyattii commented on tests/unit/util.py:
```diff
@@ -20,8 +20,8 @@ def bf16_required_version_check():
-            NCCL_MAJOR > 2 or
-            (NCCL_MAJOR == 2 and NCCL_MINOR >= 10)) and torch_info['bf16_support']:
+            NCCL_MAJOR > 2 or (NCCL_MAJOR == 2 and NCCL_MINOR >= 10)
+        ) and torch.cuda.is_available() and torch.cuda.is_bf16_supported():
```
Having a call to `torch.cuda.is_bf16_supported()` causes a problem with our unit tests (specifically the `distributed_test` decorator) because this function call initializes CUDA

mrwyattii commented on setup.py:
```diff
@@ -235,13 +234,10 @@ def create_dir_symlink(src, dest):
     else:
         nccl_version = ".".join(map(str, torch.cuda.nccl.version()[:2]))
-    if hasattr(torch.cuda, 'is_bf16_supported'):
-        bf16_support = torch.cuda.is_bf16_supported()
```
The `bf16_support` value is only used in unit testing - perhaps we could add a try...except block around this?

aphedges replied:
My first commit would be sufficient to prevent the crash because it checks whether CUDA is available.

mrwyattii replied:
The bf16 support is only utilized in unit tests - so it shouldn't affect users that build/run on different hardware

aphedges replied:
I know it's set up that way in CI, but if a developer runs the unit tests on a different machine than the build, it could still cause problems.

mrwyattii reviewed (changes requested):
It is not necessary to have `bf16_support` defined in setup.py, but we cannot have it defined during unit testing either (see the error in[ torch-latest unit tests](https://github.com/microsoft/DeepSpeed/runs/7085395930?check_suite_focus=true))

aphedges commented:
> It is not necessary to have `bf16_support` defined in setup.py, but we cannot have it defined during unit testing either (see the error in[ torch-latest unit tests](https://github.com/microsoft/DeepSpeed/runs/7085395930?check_suite_focus=true))

Are you sure the CI failure is related to my changes? I can see the same error in other PRs, such as <https://github.com/microsoft/DeepSpeed/runs/7096559886?check_suite_focus=true> for #2064.

mrwyattii commented:
> > It is not necessary to have `bf16_support` defined in setup.py, but we cannot have it defined during unit testing either (see the error in[ torch-latest unit tests](https://github.com/microsoft/DeepSpeed/runs/7085395930?check_suite_focus=true))
> 
> Are you sure the CI failure is related to my changes? I can see the same error in other PRs, such as https://github.com/microsoft/DeepSpeed/runs/7096559886?check_suite_focus=true for #2064.

You're right the current failures we're seeing on torch-latest is due to torch releasing 1.12 recently and it doesn't play well with our `distributed_test` decorator. We've disabled this workflow for now. However, please take a look at #1987 and #1990 where we moved bf16 support checking to `setup.py` for the reason of it causing issues during unit tests. We also added the torch-latest runner in #1990 as the tests that utilize bf16 are skipped on other runners.

aphedges commented:
> You're right the current failures we're seeing on torch-latest is due to torch releasing 1.12 recently and it doesn't play well with our `distributed_test` decorator. We've disabled this workflow for now. However, please take a look at #1987 and #1990 where we moved bf16 support checking to `setup.py` for the reason of it causing issues during unit tests. We also added the torch-latest runner in #1990 as the tests that utilize bf16 are skipped on other runners.

I hadn't read the explanation on #1990. Thanks for pointing it out!

As much I disagree that putting it in `setup.py` is the right place, I care the most about being able to compile `deepspeed` properly. Would you be fine with this PR if I drop my second commit, or do you want any other changes?

mrwyattii commented:
> As much I disagree that putting it in `setup.py` is the right place, I care the most about being able to compile `deepspeed` properly. Would you be fine with this PR if I drop my second commit, or do you want any other changes?

Yes, after dropping the second commit this should be good to go. And I agree that we should check for bf16 support (and perhaps nccl version too) at runtime - but we need to first resolve the issue we're having with `distributed_test` decorator. I think given that it's only used in unit tests, it's ok for now.

aphedges commented:
The failure in <https://github.com/microsoft/DeepSpeed/runs/7118253963?check_suite_focus=true> seems like a flaky performance test. DeepSpeed inference was around 10% slower than expected.

aphedges commented:
@mrwyattii, I force-pushed, and the checks passed this time. Can this PR be approved and merged?
